src/Compilers/CSharp/Portable/Binder/UnionMatchingRewriter.cs

```diff
@@ -0,0 +1,449 @@
+            Debug.Assert(result != pattern);
+            return RewritePatternWithUnionMatchingToPropertyPattern((BoundPattern)result);
+        }
+
+        protected override BoundNode? VisitExpressionOrPatternWithoutStackGuard(BoundNode node)
+        {
+            return Visit(node);
+        }
+
+        private NamedTypeSymbol ObjectType => _compilation.GetSpecialType(SpecialType.System_Object);
+
+        private static BoundPatternWithUnionMatching CreatePatternWithUnionMatching(NamedTypeSymbol inputUnionType, BoundPattern innerPattern)
+        {
+            Debug.Assert(inputUnionType.IsUnionTypeNoUseSiteDiagnostics);
+            Debug.Assert(innerPattern.InputType.IsObjectType());
+
+            var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+            PropertySymbol? valueProperty = Binder.GetUnionTypeValueProperty(inputUnionType, ref discardedUseSiteInfo);
+
+            var member = new BoundPropertySubpatternMember(innerPattern.Syntax, receiver: null, valueProperty, type: innerPattern.InputType, hasErrors: valueProperty is null).MakeCompilerGenerated();
+
+            return new BoundPatternWithUnionMatching(
+                syntax: innerPattern.Syntax,
+                inputUnionType,
+                member,
+                innerPattern,
+                inputType: inputUnionType).MakeCompilerGenerated();
+        }
+
+        public override BoundNode? VisitConstantPattern(BoundConstantPattern node)
+        {
+            node = (BoundConstantPattern)base.VisitConstantPattern(node)!;
+            if (node.IsUnionMatching)
+            {
+                return CreatePatternWithUnionMatching(
+                    (NamedTypeSymbol)node.InputType,
+                    node.Update(node.Value, node.ConstantValue, isUnionMatching: false, inputType: ObjectType, narrowedType: node.NarrowedType));
+            }
+
+            return node;
+        }
+
+        public override BoundNode? VisitRecursivePattern(BoundRecursivePattern node)
+        {
+            node = (BoundRecursivePattern)base.VisitRecursivePattern(node)!;
+            if (node.IsUnionMatching)
+            {
+                return CreatePatternWithUnionMatching(
+                    (NamedTypeSymbol)node.InputType,
+                    node.Update(
+                        node.DeclaredType, node.DeconstructMethod, node.Deconstruction, node.Properties, node.IsExplicitNotNullTest, node.Variable, node.VariableAccess,
+                        isUnionMatching: false, inputType: ObjectType, narrowedType: node.NarrowedType));
+            }
+
+            return node;
+        }
+
+        public override BoundNode? VisitListPattern(BoundListPattern node)
+        {
+            Symbol? variable = node.Variable;
+            ImmutableArray<BoundPattern> subpatterns = this.VisitList(node.Subpatterns).SelectAsArray(p => RewritePatternWithUnionMatchingToPropertyPattern(p));
```

Comment:
Alternatively, it can be rewritten to just be `.SelectAsArray(RewritePatternWithUnionMatchingToPropertyPattern)`.
```suggestion
            ImmutableArray<BoundPattern> subpatterns = this.VisitList(node.Subpatterns).SelectAsArray(static p => RewritePatternWithUnionMatchingToPropertyPattern(p));
```